### ISSUE TITLE
Publish TCK 3.1.3 (latest) to maven staging repository

### ISSUE THREAD

**Issue by alwin-joseph** [jakartaee/rest#1142]
The latest EPL licensed REST TCK (3.1.3) is available at https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee10/promoted/epl/restful-ws-tck-3.1.3.zip.

It would be convenient to get this bundle published to a central maven repository (maybe at staging https://jakarta.oss.sonatype.org/content/groups/staging/jakarta/ws/rs/ for now) so the  implementations can make use of it for running the tests.

IMHO the upload of tck jar is not recommended as the LICENCE file is within the zip file. Hence the zip file should be uploaded directly. 

I have created a Jenkins job at https://ci.eclipse.org/jakartaee-tck/job/10/job/REST_TCK_epl_staging that could be possibly used for the same but it is required to run the job from  https://ci.eclipse.org/rest. 

Can one of the committers please help publish the latest EPL TCK.

@spericas @jansupol

**Comment by spericas** [comment]
I can certainly help @alwin-joseph. I fully support the idea of having the TCKs available in Maven Central.

**Comment by jansupol** [comment]
1. There can certainly be in the maven central, the ZIP, and the JAR.
2. For it is not possible to put the ZIP file into maven central with the EFTL (Eclipse rules, did they change?), and only the EPL can be used in the ZIP in the central, and the EPL cannot be used for the vendor implementation to be certified, I do not see much difference between the ZIP and the JAR. The JAR is easier to consume and we can see the jars for other Jakarta projects (JSON-B).
3. Currently, it seems the Jakarta REST committers do not have privileges to create a new job in the Eclipse CI (at least I do not have it and a couple of other committers do not have it either). @spericas Do you have the right to create a new job in the CI?

**Comment by alwin-joseph** [comment]
>     2. For it is not possible to put the ZIP file into maven central with the EFTL (Eclipse rules, did they change?), and only the EPL can be used in the ZIP in the central, and the EPL cannot be used for the vendor implementation to be certified, I do not see much difference between the ZIP and the JAR. The JAR is easier to consume and we can see the jars for other Jakarta projects (JSON-B).
> 

Is it allowed to publish the jar file separately as it does not have the licence file. I thought not. The jar, licence and docs are bundled together in the zip ?

**Comment by waynebeaton** [comment]
> For it is not possible to put the ZIP file into maven central with the EFTL (Eclipse rules, did they change?)

It would be handy if you could cite the rule. There may be some subtlety that I'm missing.

By ZIP file, I assume that you mean the sources.

I may be splitting hairs, but it's not so much a rule imposed by the Eclipse Foundation, as it is that the EFTL applies only to the binary artifact. That is, the sources are specifically _not_ distributed under the EFTL. I can explain how all this works in painful detail if anybody is interested. 

I believe that the good folks at Maven Central will agree that it is completely valid to publish the binary artifact under the EFTL, along with a source artifact under the project license.

**Comment by alwin-joseph** [comment]
@waynebeaton 

The latest EFTL TCK bundle (zip file) released at https://download.eclipse.org/jakartaee/restful-ws/3.1/ has below contents. 
jakarta-restful-ws-tck-3.1.3.zip
- LICENSE.md (EFTL)
- jakarta-restful-ws-tck-3.1.3.jar (binary)
- docs/
- jakarta-restful-ws-tck-3.1.3.pom
- jakarta-restful-ws-tck-3.1.3-sources.jar 


1. Can we upload the jakarta-restful-ws-tck-3.1.3.zip (contains : binary + EFTL License.md + sources jar + docs) to maven ?

2. Can we upload the jakarta-restful-ws-tck-3.1.3.jar (binary with class files) separately (without any LICENSE) to maven by extracting it from the above zip bundle ?  Can we use this binary jakarta-restful-ws-tck-3.1.3.jar for the vendor implementation to be certified ?


From your email I already got clarified on the EPL files. 
```
> Can the EPL Licensed TCK archive [4] be published to maven[1] ?
Yes. 

> Can EPL TCK be used for the vendor implementation to be certified ?
No. 
```

**Comment by alwin-joseph** [comment]
> @waynebeaton
> 
> The latest EFTL TCK bundle (zip file) released at https://download.eclipse.org/jakartaee/restful-ws/3.1/ has below contents. jakarta-restful-ws-tck-3.1.3.zip
> 
>     * LICENSE.md (EFTL)
> 
>     * jakarta-restful-ws-tck-3.1.3.jar (binary)
> 
>     * docs/
> 
>     * jakarta-restful-ws-tck-3.1.3.pom
> 
>     * jakarta-restful-ws-tck-3.1.3-sources.jar
> 
> 
>     1. Can we upload the jakarta-restful-ws-tck-3.1.3.zip (contains : binary + EFTL License.md + sources jar + docs) to maven ?
> 
>     2. Can we upload the jakarta-restful-ws-tck-3.1.3.jar (binary with class files) separately (without any LICENSE) to maven by extracting it from the above zip bundle ?  Can we use this binary jakarta-restful-ws-tck-3.1.3.jar for the vendor implementation to be certified ?
> 
> 
> From your email I already got clarified on the EPL files.
> 
> ```
> > Can the EPL Licensed TCK archive [4] be published to maven[1] ?
> Yes. 
> 
> > Can EPL TCK be used for the vendor implementation to be certified ?
> No. 
> ```

@ivargrimstad  Would you have any comments on this, please.

**Comment by waynebeaton** [comment]
Sorry for the delay. 

>     1. Can we upload the jakarta-restful-ws-tck-3.1.3.zip (contains : binary + EFTL License.md + sources jar + docs) to maven ?

I'm not sure whether or not this makes any sense from a Maven perspective, but if the Maven folks are okay with it, it's okay from an EF processes POV.

>     2. Can we upload the jakarta-restful-ws-tck-3.1.3.jar (binary with class files) separately (without any LICENSE) to maven by extracting it from the above zip bundle ?  If yes, Can we use this binary jakarta-restful-ws-tck-3.1.3.jar uploaded in maven for the vendor implementation to be certified ?

Why doesn't it contain the license? Put another way... please include the license. Everything that you upload to Maven Central should have a clear indication of the license.

If it's the version of the TCK that was included in the ratification of the Final Specification, and distributed under the terms of the EFTL, then it can be used by a vendor to _self certify_ a compatible implementation.

**Comment by alwin-joseph** [comment]
> Sorry for the delay.
> 
> > ```
> > 1. Can we upload the jakarta-restful-ws-tck-3.1.3.zip (contains : binary + EFTL License.md + sources jar + docs) to maven ?
> > ```
> 
> I'm not sure whether or not this makes any sense from a Maven perspective, but if the Maven folks are okay with it, it's okay from an EF processes POV.

Thanks! 
> 
> > ```
> > 2. Can we upload the jakarta-restful-ws-tck-3.1.3.jar (binary with class files) separately (without any LICENSE) to maven by extracting it from the above zip bundle ?  If yes, Can we use this binary jakarta-restful-ws-tck-3.1.3.jar uploaded in maven for the vendor implementation to be certified ?
> > ```
> 
> Why doesn't it contain the license? Put another way... please include the license. Everything that you upload to Maven Central should have a clear indication of the license.

We only generate and release the TCK as zip bundle which contains the jar + LICENCE + others. We have never before released the jar alone hence the jar does not have the LICENCE within it.

> 
> If it's the version of the TCK that was included in the ratification of the Final Specification, and distributed under the terms of the EFTL, then it can be used by a vendor to _self certify_ a compatible implementation.

Can the LICENCE file exist outside the jar file also, but in the same maven directory as jar ? 
i.e. Can we upload jakarta-restful-ws-tck-3.1.3.jar and EFTL LICENCE.md in maven folder say for eg. jakarta/ws/rs/jakarta.ws.rs.rest-tck/3.1.3/jakarta-restful-ws-tck-3.1.3.jar and jakarta/ws/rs/jakarta.ws.rs.rest-tck/3.1.3/LICENCE.md ?

**Comment by alwin-joseph** [comment]
- I hope this issue can be closed now considering that the EFTL TCK bundle (3.1.3) is now available in the maven staging repository (https://jakarta.oss.sonatype.org/content/groups/staging/jakarta/ws/rs/jakarta-restful-ws-tck/3.1.3/ ).
- The tck jar cannot be uploaded to maven since the license is not present in it ( since the tck jar was earlier generated to be bundled along with documentation and licence files in a zip file ).

Will close the issue if no other comments.

cc @spericas @jansupol @m0mus

**Comment by mkarg** [comment]
> * The tck jar cannot be uploaded to maven since the license is not present in it ( since the tck jar was earlier generated to be bundled along with documentation and licence files in a zip file ).

Maybe I missed something, but I cannot find a confirmation on this in this thread.

**Comment by alwin-joseph** [comment]
> > * The tck jar cannot be uploaded to maven since the license is not present in it ( since the tck jar was earlier generated to be bundled along with documentation and licence files in a zip file ).
> 
> Maybe I missed something, but I cannot find a confirmation on this in this thread.

@waynebeaton did mention:

> Why doesn't it contain the license? Put another way... please include the license. Everything that you upload to Maven Central should have a clear indication of the license.

We do not have licence within the jar because the license was planned to be inside the TCK zip bundle only. Only way to put that is to have a new release!

**Comment by mkarg** [comment]
> > Why doesn't it contain the license? Put another way... please include the license. Everything that you upload to Maven Central should have a clear indication of the license.
> 
> We do not have licence within the jar because the license was planned to be inside the TCK zip bundle only. Only way to put that is to have a new release!

Why not simply putting the LICENSE file into the JAR and upload it using a new patch level version (3.1.3-1). We can publish path level changes without any formal release process.

**Comment by spericas** [comment]
We could do a patch release to include the license if that helps.

**Comment by alwin-joseph** [comment]
> Why not simply putting the LICENSE file into the JAR and upload it using a new patch level version (3.1.3-1). We can publish patch level changes without any formal release process.

- Does that mean that the jar/zip need not be released to https://download.eclipse.org/jakartaee/restful-ws/3.1/ and updated at https://github.com/alwin-joseph/specifications/blob/master/restful-ws/3.1/_index.md 
- Can we simply create the jar alone for the purpose of publishing to maven instead of the complete zip bundle.
- I see in batch TCK at https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/batch/com.ibm.jbatch.tck/2.1.1/com.ibm.jbatch.tck-2.1.1.jar the bundle license link is provided in jar manifest.mf file. Can we also consider this approach of updating the manifest.mf file with https://www.eclipse.org/legal/tck.php (EFTL). 
- Can the change https://github.com/jakartaee/rest/pull/1146 be added into this patch release ?

**Comment by mkarg** [comment]
@alwin-joseph A patch-level release (i. e. appending `-1` to the `3.1.3` version number) simply would replace the "broken" file by a "correct" one and push it on Maven Central, but *nothing* else. It is intended for cases like this, where we neither add any feature nor fix any bug. Hence, if you *want* anything *more* than that, the release process *must* be followed. But only in *that* case.

**Comment by alwin-joseph** [comment]
> @alwin-joseph A patch-level release (i. e. appending `-1` to the `3.1.3` version number) simply would replace the "broken" file by a "correct" one and push it on Maven Central, but _nothing_ else. It is intended for cases like this, where we neither add any feature nor fix any bug. Hence, if you _want_ anything _more_ than that, the release process _must_ be followed. But only in _that_ case.


As suggested, I created a patch release of the TCK, as version 3.1.3-1 at https://www.eclipse.org/downloads/download.php?file=/ee4j/jakartaee-tck/jakartaee10/staged/eftl/jakarta-restful-ws-tck-3.1.3-1.zip where the jar file (jakarta-restful-ws-tck-3.1.3-1.jar) has the EFTL LICENSE file.

Also I have published the same jar to maven staging at https://jakarta.oss.sonatype.org/content/groups/staging/jakarta/ws/rs/jakarta-restful-ws-tck/3.1.3-1/jakarta-restful-ws-tck-3.1.3-1.jar.

Comments are welcome.

**Comment by rzo1** [comment]
Any plan for getting this artifact to central? Would be very convinient for inclusion ;-)

**Comment by jamezp** [comment]
@alwin-joseph Is there a reason it doesn't show up here https://download.eclipse.org/jakartaee/restful-ws/3.1/? I was looking for the sha1 and a more direct link :)

FWIW this passes with RESTEasy and WildFly.

**Comment by spericas** [comment]
@alwin-joseph ?

**Comment by alwin-joseph** [comment]
Sure, I will work to publish the TCK jar 3.1.3-1 from staging (https://jakarta.oss.sonatype.org/content/groups/staging/jakarta/ws/rs/jakarta-restful-ws-tck/3.1.3-1/) to maven central. 

> @alwin-joseph Is there a reason it doesn't show up here https://download.eclipse.org/jakartaee/restful-ws/3.1/? I was looking for the sha1 and a more direct link :)
> 

Hi @mkarg 

Can the patch level release (3.1.3-1) be uploaded to  https://download.eclipse.org/jakartaee/restful-ws/3.1 ? The only change is that it has the EFTL LICENCE file within the TCK jar also.

cc @ivargrimstad 

> FWIW this passes with RESTEasy and WildFly.

Yes the 3.1.3-1 TCK is same as 3.1.3 except that the EFTL LICENCE file was added to the jar in 3.1.3-1.

**Comment by alwin-joseph** [comment]
> Sure, I will work to publish the TCK jar 3.1.3-1 from staging (https://jakarta.oss.sonatype.org/content/groups/staging/jakarta/ws/rs/jakarta-restful-ws-tck/3.1.3-1/) to maven central.

The TCK jar 3.1.3-1 was uploaded to maven central at https://mvnrepository.com/artifact/jakarta.ws.rs/jakarta-restful-ws-tck. 

Service release 3.1.4 of the TCK will be available ( ref https://github.com/jakartaee/specifications/pull/657).
I have asked whether the maven location of the new TCK jar can be updated in https://download.eclipse.org/jakartaee/restful-ws/3.1. https://github.com/jakartaee/specifications/pull/657#issue-1926170954.

**Comment by jamezp** [comment]
Should this be closed now that 3.1.4 has been released?
